nabs/callbacks.py

```diff
@@ -0,0 +1,166 @@
+                  'float': '{:.2f}',
+                  'number': '{:.3g}',
+                  'integer': '{:d}'}
+
+    _html_head = """<!DOCTYPE html>
+<html>
+<head>
+<style>
+table {
+  font-family: arial, sans-serif;
+  border-collapse: collapse;
+  width: 100%;
+}
+
+td, th {
+  border: 1px solid #dddddd;
+  text-align: left;
+  padding: 8px;
+}
+
+tr:nth-child(even) {
+  background-color: #dddddd;
+}
+</style>
+</head>
+<body>
+"""
+
+    _html_tail = "</body> </html>"
+
+    def __init__(self, elog, ipython):
+        self._elog = elog
+        self._ipython = ipython
+        self._data = dict()
+        self._descriptors = set()
+        self._data_keys = []
+        self._format_info = dict()
+
+        # TO-DO:
+        # - add granular switches as optional arguments
+        # - add error handling for misspelled keys
+        # - add default lists for start doc keys, elog tags?
+        # - handling if data key collides with seq_num, time?
+        # - consider pulling precision from data
+
+    def start(self, doc):
+        """ Post plan information on start document"""
+        self._send_post = doc.get('post', self._elog.enable_run_posts)
+
+        # Clean up before starting new run
+        self._data = dict()
+        self._descriptors = set()
+        self._data_keys = []
+        self._format_info = dict()
+
+        if self._send_post:
+            # Grab last ipython input
+            run_info = self._ipython.user_ns["In"][-1]
+            logger.info("Posting run start information to elog")
+            self._elog.post(run_info, tags=['plan_info', 'RE'])
+
+        super().start(doc)
+
+    def descriptor(self, doc):
+        """ Initialize table information """
+        # these always exist
+        self._data = dict(seq_num=[], time=[])
+        for k in doc['data_keys']:
+            dk_entry = doc['data_keys'][k]
+
+            if dk_entry['dtype'] not in self._FMTLOOKUP:
+                logger.warn(f'skipping {k}, format not recognized')
+                continue
+
+            self._data.update({k: []})
+            self._data_keys.append(k)
+            self._format_info.update({k: dk_entry['dtype']})
+
+        self._descriptors.add(doc['uid'])
+        # To-do: consider signals with mismatched keys
+
+    def event(self, doc):
+        if doc['descriptor'] not in self._descriptors:
+            return
+
+        self._data['seq_num'].append(doc['seq_num'])
+        self._data['time'].append(
+            str(datetime.fromtimestamp(doc['time']).time())[:-4]
+            )
+
+        # this might break for unfilled data, but do we deal with that?
+        for k in self._data_keys:
+            try:
+                dtype = self._format_info[k]
+                style = self._FMTLOOKUP[dtype]
+                self._data[k].append(style.format(doc['data'][k]))
+            except Exception as ex:
+                print(ex)
+                logger.warning(f'key: {k} not in event document')
```

Comment:
Could be a different error, like `style` failing to format too

Also, should `self._data[k]` get a value in this scenario, just in case? Or will remaining data be off-by-one?

Reply:
Could you clarify what you mean by "get a value"?  Do you mean be supplied a default value to keep the table aligned?

Reply:
Yes, keeping the table aligned with some default value